### ISSUE TITLE
Separate compile and runtime dependencies of external modules

### ISSUE THREAD

**Issue by NHendricks** [gradle/gradle#977]
If I declare a compile dependency (e.g. to io.reactivex:rxnetty:0.4.19) I see that gradle also puts all transitive runtime dependencies into the compile configuration (in this case for example io.reactivex:rxjava:1.0.17). 
### Expected Behavior

Example:
`dependencies {compile 'io.reactivex:rxnetty:0.4.19'}`
My Java Code using the class `rx.exceptions.Exceptions` should not compile.
The compile should fail because the dependency io.reactivex:rxnetty:0.4.19 provides only runtime dependencies and rx.exceptions.Exceptions is part of io.reactivex:rxjava.

### Current Behavior

The transitive runtime dependencies are added to the compile configuration. They should not be added to the compile configuration.

### Context

It leads to unwanted dependendencies and has a reasonable impact for the migration to gradle in our company regarding tool acceptance and management of our dependencies. 

### Steps to Reproduce (for bugs)

I will upload a sample project here.

### Your Environment

There is no specific environment needed. A pure call of gradle dependencies in the sample project shows the tree including the transitive runtime dependencies in the compile configuration.
[gradle_example.zip](https://github.com/gradle/gradle/files/634150/gradle_example.zip)

 * Build scan URL:

**Comment by oehme** [comment]
We are actively working on this, so you can expect a fix in Gradle 3.4.

**Comment by oehme** [comment]
This is fixed for project dependencies in 3.4 if you use the new java-library plugin. We still need to fix it for external dependencies (e.g. separating `compile` and `runtime` scope for Maven dependencies).

**Comment by eriwen** [comment]
@oehme Is this still scheduled for 4.0?

**Comment by oehme** [comment]
I don't think we'll be able to make it. 4.1 seems more realistic.

**Comment by eriwen** [comment]
@oehme cool, moved

**Comment by CoderSpinoza** [comment]
Is this going to be scheduled soon? :)

**Comment by jjohannes** [comment]
This is now implemented (see #3360). Please feel free to give the latest nightly a try.

It will be in `4.5`, but we will only make it the default behavior in 5.0 as it might break existing builds.

If you try the nightly, you can switch the behavior on by adding this to your `settings.gradle`:
`gradle.experimentalFeatures.enable()`

(4.5 will include this opt-in switch as well but the syntax might change for the release)

**Comment by julioromano** [comment]
What is the syntax of the opt in switch in 4.5-rc-2 ?

**Comment by jjohannes** [comment]
In the end, we did not make any changes for 4.5. It's still `gradle.experimentalFeatures.enable()`. 

4.6 will include an "official" way to enable this. I updated the comment above.

**Comment by oehme** [comment]
Closing the loop here, Gradle 4.6 added 

`enableFeaturePreview('IMPROVED_POM_SUPPORT')` in settings.gradle

for this. Gradle 5.0 will make it the default.

**Comment by robstoll** [comment]
@oehme you mentioned in feb 2017
> We still need to fix it for external dependencies (e.g. separating compile and runtime scope for Maven dependencies).

Is this fixed by now and if so, what do I need to do that maven `<scope>runtime</scope>` dependencies are not treated as `compile`?

**Comment by oehme** [comment]
This is for external dependencies, see the title of the issue.

**Comment by robstoll** [comment]
Well then, it still does not work with 4.5.1
I got a maven project which has a dependency defined as `<scope>runtime</scope>`, I have put `gradle.experimentalFeatures.enable()` in my gradle.properties. I am depending on the maven project and can still access code from the dependency defined in scope `runtime`.

**Comment by robstoll** [comment]
Man... my fault, I should have put it in settings.gradle => works. Sorry for the noise

**Comment by oehme** [comment]
@trevjonez Sorry, I don't see the problem. The compile classpath contains your api dependency and the runtime classpath contains your implementation dependency, just as expected.

**Comment by trevjonez** [comment]
@oehme I didn't do a good job of explaining it. I am expecting debugCompileClasspath to resolve as:
```
\--- com.trevjonez:and-lib:0.1.0
     \--- com.trevjonez:and-lib_release:0.1.0
          \--- io.reactivex.rxjava2:rxjava:2.2.0
               \--- org.reactivestreams:reactive-streams:1.0.2
```

but it is resolving as:
```
\--- com.trevjonez:and-lib:0.1.0
     \--- com.trevjonez:and-lib_release:0.1.0
          +--- io.reactivex.rxjava2:rxjava:2.2.0
          |    \--- org.reactivestreams:reactive-streams:1.0.2
          \--- com.squareup.moshi:moshi:1.6.0
               \--- com.squareup.okio:okio:1.14.0
```

all of the implementation dependencies are being exposed on the compile classpath as well as the runtime classpath.

**Comment by oehme** [comment]
Are you sure you linked the right test case? I only see asserts on files being there.

**Comment by trevjonez** [comment]
oops, I linked to master and had removed the asserts last night. I updated the original link [with this commit specific url](https://github.com/trevjonez/android-components-plugin/blob/227280053cd3ecf8a8899d7ecd260c4a78f20831/plugin/src/test/kotlin/com/trevjonez/AndroidComponentsPluginTest.kt#L135)

**Comment by trevjonez** [comment]
@oehme 
here is a build scan for the library publishing build:
https://scans.gradle.com/s/3x4vcfaxtgt5m/

here is a build scan for the consuming application build: https://scans.gradle.com/s/mfmp2gtq2duwk/dependencies?toggled=W1swXSxbMCw3XSxbMCw3LFswXV0sWzAsNyxbMCwxXV0sWzAsNyxbMCwxLDRdXSxbMCw3LFswLDEsMl1dLFswLDhdLFswLDgsWzBdXSxbMCw4LFswLDFdXSxbMCw4LFswLDEsNF1dLFswLDgsWzAsMSwyXV1d

**Comment by trevjonez** [comment]
so I went that route so that I could still support consumers depending on the library without specifying the default variant artifact directly. I opened this issue explaining the issue I was seeing when publishing the same artifact under both the root component and release variant. https://github.com/gradle/gradle/issues/6389

**Comment by oehme** [comment]
Have you checked whether the POM for that default variant actually has properly separated `<compile>/<runtime>` dependencies? Just making sure this is not a problem on the publishing side.

**Comment by trevjonez** [comment]
I did yea. Here is a gist with the relevant POM files and the .module so you can see what is placed there in order for me to generate a usable POM given the issues from 6389.

https://gist.github.com/trevjonez/62b3d7907c13ca0728139d85ddf9cfbb

**Comment by oehme** [comment]
Oh, you simply forgot the `IMPROVED_POM_SUPPORT` feature preview, that's what the api/impl separation is bound to until Gradle 5.0

**Comment by oehme** [comment]
Can you try consuming that lib from a plain Java project? Just making sure this issue is not specific to the Android plugin.

**Comment by trevjonez** [comment]
looks like a java library is also pulling in all the runtime transitives as compile dependencies

https://scans.gradle.com/s/dmsu7idjx2c2k/dependencies?toggled=W1swXSxbMCwxXSxbMCwxLFswXV0sWzAsMSxbMCwxXV0sWzAsMSxbMCwxLDRdXSxbMCwxLFswLDEsMl1dXQ

scan from java-lib project after modifications in [this commit](https://github.com/trevjonez/android-components-plugin/commit/f900b6c6bf38b53c8b36921534868f717508db8b)

**Comment by bigdaz** [comment]
It's exactly as you suspected @oehme: when the packaging is not a ["known Jar packaging"](https://github.com/gradle/gradle/blob/440e69376cee7492da15284e2b78701249e2b6c1/subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/maven/DefaultMavenModuleResolveMetadata.java#L55) then we don't create any variants for the module, which means that the compile/runtime separation is lost.

The non-jar packaging case is not one I considered when I put together [this complicated sheet of dependency interactions](https://docs.google.com/spreadsheets/d/1PC7pbZqxjvYBh4A85398GTDLHbniwt14XqCT97gfdIw/edit#gid=0). (Apologies, I can't seem to make that sheet public). We really need to think about the behaviour we want to make the default in 5.0.

**Comment by oehme** [comment]
Thanks Daz, I've opened a follow-up issue for 5.0

**Comment by Ethan1983** [comment]
I can confirm this is fixed from Gradle 5.0 for transitive aar (android) runtime dependency.
